core/src/rpc/aggregator.rs

```diff
@@ -393,116 +394,129 @@ impl ClementineAggregator for Aggregator {
-                }
+            let verifier_public_keys = verifier_public_keys.clone();
+            async move {
+                let response = client
+                    .set_verifiers(Request::new(verifier_public_keys))
+                    .await?;
+                Ok::<_, Status>(response.into_inner())
             }
         }))
         .await?;
 
         tracing::info!("Collecting operator details...");
-        let operator_params = try_join_all(self.operator_clients.iter().map(|client| {
-            let mut client = client.clone();
-            async move {
-                let mut responses = Vec::new();
-                let mut params_stream = client
-                    .get_params(Request::new(Empty {}))
-                    .await?
-                    .into_inner();
-                while let Some(response) = params_stream.message().await? {
-                    responses.push(response);
+        let mut operator_params_response_streams =
+            try_join_all(self.operator_clients.iter().map(|client| {
+                let mut client = client.clone();
+                async move {
+                    let stream = client
+                        .get_params(Request::new(Empty {}))
+                        .await?
+                        .into_inner();
+                    Ok::<_, Status>(stream)
                 }
+            }))
+            .await?;
 
-                Ok::<_, Status>(responses)
-            }
-        }))
-        .await?;
+        let mut set_operator_request_streams =
+            try_join_all(self.verifier_clients.iter().map(|client| {
+                let mut client = client.clone();
+                async move {
+                    let (tx, rx) = channel(1280);
+                    let stream = client.set_operator(ReceiverStream::new(rx)).await?;
+                    Ok::<_, Status>((stream.into_inner(), tx))
+                }
+            }))
+            .await?;
 
         tracing::info!("Informing verifiers for existing operators...");
-        try_join_all(self.verifier_clients.iter().map(|client| {
-            let mut client = client.clone();
-            let operator_params = operator_params.clone();
-
-            async move {
-                for params in operator_params {
-                    let (tx, rx) = tokio::sync::mpsc::channel(1280);
-                    let future =
-                        client.set_operator(tokio_stream::wrappers::ReceiverStream::new(rx));
-
-                    for param in params {
-                        tx.send(param)
-                            .await
-                            .map_err(|_| output_stream_ended_prematurely())?;
-                    }
-
-                    future.await?; // TODO: This is dangerous: If channel size becomes not sufficient, this will block forever.
+        let operator_params_setup_task_handle = tokio::spawn(async move {
+            for response in operator_params_response_streams.iter_mut() {
```

Comment:
Currently this loop:
1. For each operator
2. Stream all messages to operator (await)

This will block for each operator and won't pipeline operators in parallel.

can we collect all async blocks on the inside and join them on the outside of this loop?